include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h

```diff
@@ -154,8 +223,54 @@ class PreconditionerML : public PreconditionerBase<double>
     // re-calculate matrix
     pde_operator.calculate_system_matrix(system_matrix);
 
-    // initialize Trilinos' AMG
-    amg.initialize(system_matrix, ml_data);
+    if(ml_operator_type == MLOperatorType::Laplace)
+    {
+      // Default setup; add near null space for Laplace operator.
+      amg.initialize(system_matrix, ml_data);
+    }
+    else if(ml_operator_type == MLOperatorType::Elasticity)
+    {
+      // get Teuchos::ParameterList to provide custom near null space basis
+      unsigned int const     dimension      = pde_operator.get_matrix_free().dimension;
+      Teuchos::ParameterList parameter_list = get_ML_parameter_list(ml_data, dimension);
+
+      std::vector<std::vector<double>> rigid_body_modes;
+
+      // If we use the AMG preconditioner in a multigrid setting,
+      // signaled by `dof_handler.has_level_dofs() == true`,
+      // we construct the AMG preconditioner on the coarsest level.
+      // Otherwise, we use AMG as fine-level preconditioner.
+      if(pde_operator.get_matrix_free().get_dof_handler().has_level_dofs())
+      {
+        rigid_body_modes = dealii::DoFTools::extract_level_rigid_body_modes(
+          0,
+          *pde_operator.get_matrix_free().get_mapping_info().mapping,
+          pde_operator.get_matrix_free().get_dof_handler(),
+          dealii::ComponentMask(dimension, true));
+      }
+      else
+      {
+        rigid_body_modes = dealii::DoFTools::extract_rigid_body_modes(
+          *pde_operator.get_matrix_free().get_mapping_info().mapping,
+          pde_operator.get_matrix_free().get_dof_handler(),
+          dealii::ComponentMask(dimension, true));
+      }
```

Comment:
It looks like the natural place for this code would be the implementation of the Elasticity operator.

Reply:
So I mean option ii) described here: https://github.com/exadg/exadg/pull/693#discussion_r1795030659

Reply:
we now have `OperatorBase::get_constant_modes()`, which creates default modes, and the derived classes then can overwrite this function as done in `ElasticityOperatorBase`.
The modes can have different types though, which is a bit of a bummer ...
there are `std::veector<std::vector<bool>> constant_modes` for cases when only identifying a component is enough (Poisson problems for example), while there are also
`std::vector<std:.vector<double>> constant_modes_values`, which is the more general case, describing modes comprised of non-trivial entries.
I think the former still exists for performance/storage reasons and backward compatibility. I decided to keep them both and the `constant_modes_values` is checked for values first, which mirrors the behavior within deal.II. On our side, we have to provide one of the two, otherwise we assert.